### ISSUE TITLE
Updating UI in angular and scope digest errors

### ISSUE THREAD

**Issue by cramatt** [apollostack/angular1-apollo#14]
I have an angular 1 app using this library with a graphql server. 

My understanding is that with Angular 1, you need to use `apollo.client.watchQuery` to get auto updating UIs, where as in Angular 2 or React you get this for free using `apollo.query`. Is this correct? 

For example I have a UI component populated using this query: 

```
apollo.query({
  query,
  variables: {
    sessionId: this.sessionId
  }
}).then(result => {
  debugger;
  this.stations = result.data.stations;
  this.instructorStations = result.data.instructorSession.associatedStations;
});
```

And another component that performs this mutation: 

```
const mutation = gql `
  mutation removeStationFromSession($removeStationFromDashboardInput:removeStationFromDashboardInput!) {
    removeStationFromDashboard(input:$removeStationFromDashboardInput) {
      InstructorSession {
        associatedStations {
          id
        }
      }
    }
  }
`;
```

Should I expect the first UI component to update in this case? 

Secondly, when I use `apollo.client.watchQuery` and subscribe to the result, I need to map the results to a plain JS object otherwise I'm getting a scope digest error. Here is my code: 

```
const observable = apollo.client.watchQuery({
  query,
  variables: {
    sessionId: this.sessionId
  }
})

observable.subscribe({
  next: result => {
    this.stations = result.data.stations.map(item => {
      return {
        name: item.name,
        id: item.id
      }
    });
    this.instructorStations = result.data.instructorSession.associatedStations.map(item => {
      return {
        name: item.name,
        id: item.id
      }
    });
    $scope.$apply();
  }
});
```

Without the mapping, I get an error that digest cycle reached maximum cycles. I suspect this is because of properties that Apollo adds to the object which Angular chokes on when doing diffs (ex: Symbol(Id)). 

Any ideas here?

**Comment by Darkbladecr** [comment]
I suggest you have a read of this post: https://dev-blog.apollodata.com/graphql-subscriptions-in-apollo-client-9a2457f015fb#.thtuze993

the way I understand it is that watchQuery is a subscription, which allows implementation of realtime updates to the UI. Now this being said it is not as simple as changing your code from `apollo.query()` to `apollo.watchQuery()`. You will need to also add a pubsub system onto your server and publish the content you want pushed to the UI following mutations. 

Also I think you mean to use `apollo.watchQuery` instead of `apollo.client.watchQuery`

The recent branch of `rxjs` is a step in the right direction, however I am still struggling to implement it correctly as there is a lack of documentation. The medium post does shed some light, but [code is missing](https://github.com/apollostack/subscriptions-transport-ws/issues/38) from the [`subscriptions-transport-ws`](https://github.com/apollostack/subscriptions-transport-ws) module.

I have added the missing `addGraphQLSubscriptions` function from the [GitHunt example](https://github.com/apollostack/GitHunt-React/blob/master/ui/helpers/subscriptions.js). However my resolve functions don't work when using a schema root of subscriptions. Otherwise when I try to link it to queries my resolver function returns undefined.

You are more than welcome to have a look at [my work so far at implementing with TodoMVC.](https://github.com/Darkbladecr/angular-apollo-todo/tree/rxjs)

**Comment by Darkbladecr** [comment]
A little update.

It seems that the AST validator is not letting me use the subscription name, so I have settled for query, as per the [apollo documentation](http://dev.apollodata.com/core/apollo-client-api.html#ApolloClient\.watchQuery).

Unfortunately my pubsub doesn't seem to be publishing the updates. Interestingly when I try to subscribe directly through my node server I am getting the `Error: Must provide schema`, which I do not understand. Not sure if you can see anything wrong with this setup:

```
// subscribe.js
import { PubSub, SubscriptionManager } from 'graphql-subscriptions';
import schema from './schema';

const pubsub = new PubSub();
const subManger = new SubscriptionManager({
	schema,
	pubsub,
	setupFunctions: {
		todoAdded: (options, args) => ({
			todoAdded: (todo) => {
				return true;
			},
		}),
	},
});

subManger.subscribe({
	query: `subscription todos($completed:Boolean) {
				todoAdded(completed:$completed){
					id
					title
					completed
				}
			}`,
	variables: {
		completed: null
	},
	context: {},
	callback: (err, data) => console.log(data)
});

pubsub.publish('todoAdded', {
	id: 123,
	title: 'testing graphql subscriptions',
	completed: false
});

export { subManger, pubsub };
```

```
// schema.js
import Resolvers from './resolvers';
import Schema from './schema.graphql';
import { makeExecutableSchema } from 'graphql-tools';

const executableSchema = makeExecutableSchema({
	typeDefs: Schema,
	resolvers: Resolvers,
	allowUndefinedInResolve: true,
	resolverValidationOptions: {
		requireResolversForNonScalar: false
	},
	printErrors: process.env.NODE_ENV !== 'production',
});

export default executableSchema;
```

```
// schema.graphql
type Todo {
	id: ID
	createdAt: String
	updatedAt: String
	title: String
	completed: Boolean
}

type User {
	id: ID
	createdAt: String
	updatedAt: String
	firstName: String
	lastName: String
}

type Query {
	todos(completed: Boolean): [Todo]
	todoAdded(completed: Boolean): Todo
}

type Mutation {
	addTodo(title: String!): Todo
	editTodo(id: Int!, title: String, completed: Boolean): Todo
	deleteTodo(id: Int!): String
}

type Subscription {
	todoAdded(completed: Boolean): Todo
}

schema {
	query: Query
	mutation: Mutation
	subscription: Subscription
}
```

**Comment by kamilkisiela** [comment]
@cramatt To avoid an error with the digest cycle reached maximum, use [this](https://github.com/wolfgangGoedel/observe-on-scope/blob/master/src/safeApplyOnScope.ts#L5). It checks if digest cycle is running.

**Comment by kamilkisiela** [comment]
Also, `watchQuery()` is for keep getting the new values that were updated in the store. It's not a subscription, subscriptions are subscriptions, that's whole a new story :) 

If you use `query()` you fetch data only once. If you use `watchQuery()` every change in the store will push new values to your `Subscription` that comes from `Observable` returned by that `watchQuery()` method. 

It means that if mutation happens and you update a query, the store is being also updated to this way you push theses changes into your `Subscription` (from `Observable`, not subscription feature).

**Comment by kamilkisiela** [comment]
@Darkbladecr Here's an example of using `observeOnScope` that I recommended in my PR (rxjs branch).

https://github.com/wolfgangGoedel/observe-on-scope/blob/master/example/component.ts#L15

```ts
const obs = this.client.watchQuery({...})
  .observeOnScope($scope)
  .subscribe(result => ... );
```

**Comment by Darkbladecr** [comment]
@kamilkisiela Do you mind posting an example with .observeOnScope() ? When I try to use it despite building and importing the observe-on-scope module I am getting errors with method not found.

**Comment by kamilkisiela** [comment]
Sure, I will work on this soon :+1:

**Comment by kamilkisiela** [comment]
@Darkbladecr okay, done
https://github.com/apollographql/angular1-apollo/tree/rxjs/examples/hello-world
